### ISSUE TITLE
Browser compatibility? / Hash issue?

### ISSUE THREAD

**Issue by kentaron04** [addrummond/newibexfarmissues#3]
Hi, 

Thanks a lot for updating Ibex Farm.  The new version seems much cleaner, and easier to use!

Two issues though.

[1]
The other day, I held a final exam online using Ibex Farm (because the randomizing function is very useful), with about 75 participants.  About 4 participants reported that they could not finish the trial: at the final stage, they saw a never-ending "Sending results" message with a rotating bar, and could not get to the completion message.  The results were not uploaded either.  It seems that most of them were using Internet Explorer, but maybe not?? (See below).

I quickly set up the same experiment (I mean, exam) in the older version of Ibex Farm, and two of them took it and could successfully finish it.  (The others took the exam using smartphones.)

The record of the old Ibex Farm indicates that the User Agents of these two test-takers were: 
# USER AGENT: Mozilla/5.0 (Windows NT 6.3; WOW64; Trident/7.0; rv:11.0) like Gecko
# USER AGENT: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36 Edg/87.0.664.75

Note that later, I held another online exam using the old version of Ibex Farm (fearing that the same problem would emerge), and the 100 plus participants could all finish the exam without errors.  

[2]
In the new Ibex Farm results data, I found that many participants were assigned identical MD5 hashes.  As far as I understand, identical hashes may be assigned to different participants if they were using the same browsers under the same IP addresses, and this occasionally happened in the past with the old Ibex Farm.  But the number of identical hashes was overwhelming this time, and I'm wondering whether this is a bug or a different calculation method is now employed for hash generation.  Could you take a look at the link below, where the hashes of all the participants were cited?  Note that when I later held another exam using the old Ibex farm as mentioned above, with 100 plus participants, every participant was assigned a different hash.  [https://www.dropbox.com/s/9uyx6ql1e7wqao3/results_excerpts2.txt?dl=0](url)

Thanks!

-Ken Nakatani

PS: I run linguistics experiments too.

**Comment by addrummond** [comment]
@kentaron04 

Hello Ken,

Thank you for this very detailed and helpful report. The information you've provided has made it vastly easier to look into these issues than it could have been!

Browser user agent strings are rather counterintuitive. The first is from IE 11; the second is from a slightly outdated version of Edge (released in Oct 2020).

[1] I've confirmed that results submission was not working on IE 11. The reason is that the code for compressing results made use of the `slice` method of `Uint8Array`, which isn't implemented in IE11. I've fixed this by adding a polyfill. I've also explicitly disabled results compression on versions of IE prior to 11. On the version of Edge that I have to hand, I can't reproduce any problem. It frankly seems unlikely that a version of Edge released in Oct 2020 would fail to work. I have to suspect that this user might have used IE the first time around and Edge the second.

[2] This is puzzling. It's true that the method for generating unique IDs has changed, but this doesn't explain what you're seeing. Previously, the ID was a hash based on a browser fingerprint. However, this raised GDPR concerns for some users, as it would be possible – theoretically – to recover some of the fingerprint data from the hash. (The hash isn't salted and there is a relatively limited number of probable fingerprints, so a brute force approach could be effective.) The hash is now based on a randomly generated number that's stored in a cookie. If the same browser is used to take an experiment multiple times, then you'll get the same hash unless the `ibex_random_id` cookie is cleared between experiments. The only explanation I can think of is this obvious one – some students were sharing the same computer(s). I haven't been able to reproduce any issues with duplicate hash generation so far.

In summary, [1] should now be fixed. Let's keep an eye on [2].

**Comment by kentaron04** [comment]
Hi Alex, 

Thanks for your prompt reply.  

Regarding [1], I’m glad to hear that you could pin down the problem.  Thanks!

Regarding [2]: 
>The only explanation I can think of is this obvious one – some students were sharing the same computer(s). 

That’s a logical possibility but that’s unlikely for the following reasons:

-This was an exam which the students had to take at the same time, so in order for them to take it with the same computers, they needed to be in the same room(s).
-However, our university does not have a dorm, and the majority of the students are living with their families, commuting to school.
-And the exam was held in the period when most classes went online, so there were not many students on campus.
-Also, for a proctoring purpose, I asked all of them to join a Zoom meeting and keep the video on while taking the exam, and from what I saw, they seem to be in their own rooms.  
-If we see the following file where the data are sorted according to hashes, we see that even some iPhone users (with different OS versions) were assigned the same hashes—it’s quite unlikely that they shared the same iPhones.

https://www.dropbox.com/s/zwzxutle0ro6h44/results_excerpts2_sorted.txt?dl=0

Note that they were asked not to start the exam until I told them to do so, and because I forgot to specify the _Set_Counter_ function, most of them (67 students out of 71) got the same counter number.  Could this be the cause of the issue?

A side note: in the second exam I held using the old Ibex Farm, all the students (100+) were assigned different hashes (even without _Set_Counter_).  

https://www.dropbox.com/s/xlkhbjwsr6io4jy/results_excepts.txt?dl=0

Thanks!

-Ken N.

**Comment by addrummond** [comment]
@kentaron04 Thanks for the additional info. The counter should not influence the value of the hash. At present I've not been able to reproduce the problem. If I submit results for an experiment multiple times (using different browsers, or clearing my cookies after each submission), then I see a different hash for each set of results. I'll continue thinking about what might have caused your results.

**Comment by kentaron04** [comment]
OK, got it, thanks a lot!!

-KN

2021年2月1日(月) 18:41 Alex Drummond <notifications@github.com>:
